src/main/java/org/tikv/txn/TxnKVClient.java

```diff
@@ -1,5 +1,363 @@
+    }
+
+    public TiConfiguration getConf() {
+        return conf;
+    }
+
+    public ReadOnlyPDClient getPdClient() {
+        return pdClient;
+    }
+
+    public TxnKVClient(TiConfiguration conf,
+                       RegionStoreClient.RegionStoreClientBuilder clientBuilder, PDClient pdClient) {
+        this.conf = conf;
+        this.clientBuilder = clientBuilder;
+        this.regionManager = clientBuilder.getRegionManager();
+        //this.session = TiSession.create(TiConfiguration.createRawDefault(addresses));
+        //this.regionManager = new RegionManager(session.getPDClient());
+        this.pdClient = pdClient;
+    }
+
+    /*public static TxnKVClient createClient(String addresses) {
+        return new TxnKVClient(addresses);
+    }*/
+
+    public TiTimestamp getTimestamp() {
+        BackOffer bo = ConcreteBackOffer.newTsoBackOff();
+        TiTimestamp timestamp = new TiTimestamp(0, 0);
+        try {
+            while(true) {
+                try {
+                    timestamp = pdClient.getTimestamp(bo);
+                    break;
+                } catch (final TiKVException e) {//retry is exhausted
+                    bo.doBackOff(BackOffFunction.BackOffFuncType.BoPDRPC, e);
+                }
+            }
+        } catch (GrpcException e1) {
+            LOG.error("Get tso from pd failed,", e1);
+        }
+        return timestamp;
+    }
+
+    /**
+     * Begin a new transaction
+     * @return
+     */
+    public ITransaction begin() {
+        return new TikvTransaction(this);
+    }
+
+    public ITransaction begin(Function<ITransaction,Boolean> function) {
+        return new TikvTransaction(this, function);
+    }
+
+    //add backoff logic when encountered region error,ErrBodyMissing, and other errors
+    public ClientRPCResult prewrite(BackOffer backOffer, List<Kvrpcpb.Mutation> mutations, byte[] primary, long lockTTL, long startTs, long regionId) {
+        ClientRPCResult result = new ClientRPCResult(true, false, null);
+        //send request
+        TiRegion region = regionManager.getRegionById(regionId);
+        RegionStoreClient client = clientBuilder.build(region);
+        try {
+            client.prewrite(backOffer,  ByteString.copyFrom(primary), mutations, startTs, lockTTL);
+        } catch (final TiKVException | StatusRuntimeException e) {
+            result.setSuccess(false);
+            result.setRetry(e instanceof RegionException);//mark retryable, region error, should retry prewrite again
+            result.setError(e.getMessage());
+        }
+        return result;
+    }
+
+    /**
+     * Commit request of 2pc,
+     * add backoff logic when encountered region error, ErrBodyMissing, and other errors
+     * @param backOffer
+     * @param keys
+     * @param startTs
+     * @param commitTs
+     * @param regionId
+     * @return
+     */
+    public ClientRPCResult commit(BackOffer backOffer, byte[][] keys, long startTs, long commitTs, long regionId) {
```

Comment:
I think as an SDK, the user needn't know 2PC,  so could we make functions `prewrite` and `commit` private?

Reply:
> I think as an SDK, the user needn't know 2PC, so could we make functions `prewrite` and `commit` private?

Maybe upper KV-engine driven encapsulation will need 2PC process to be customized...